File: app/viewport/camera.py
"""Orbit camera for the viewport.

Owns the orbit state (target/distance/yaw/pitch/focal), the working-scale
envelope, framing, and screen rays — every place the camera math or its
scale floors live. The widget owns policy (when to reframe, animation
timing); the camera owns mechanism. `view_scale` is meters per working
unit: floors and ceilings scale with it so mm and km work stay reachable
while the grid remains a truthful world-space snap reference.
"""
from __future__ import annotations

import math

import numpy as np

# The familiar startup framing: 6 m away from a 1 m grid. Working-unit
# switches reproduce this view scaled by the unit factor.
DEFAULT_VIEW_DISTANCE = 6.0
_WORLD_UP = np.array([0.0, 0.0, 1.0])
_FALLBACK_UP = np.array([0.0, 1.0, 0.0])


class ViewportCamera:
    def __init__(self) -> None:
        self.target = np.array([0.0, 0.0, 0.0], dtype=np.float64)
        self.distance = DEFAULT_VIEW_DISTANCE
        self.yaw = math.radians(35.0)
        self.pitch = math.radians(28.0)
        self.focal = 1.5
        self.view_scale = 1.0  # meters per working unit

    # -- basis ---------------------------------------------------------------

    def position(self) -> np.ndarray:
        cos_pitch = math.cos(self.pitch)
        offset = np.array([
            cos_pitch * math.cos(self.yaw),
            cos_pitch * math.sin(self.yaw),
            math.sin(self.pitch),
        ])
        return self.target + self.distance * offset

    def basis(self) -> tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
        """(position, forward, right, up), matching the shader convention."""
        position = self.position()
        forward = self.target - position
        forward = forward / max(np.linalg.norm(forward), 1e-9)
        world_up = _WORLD_UP
        if abs(np.dot(forward, world_up)) > 0.99:
            world_up = _FALLBACK_UP
        right = np.cross(forward, world_up)
        right /= max(np.linalg.norm(right), 1e-9)
        up = np.cross(right, forward)
        return position, forward, right, up

    # -- interaction -----------------------------------------------------------

    def orbit(self, delta_x: float, delta_y: float) -> None:
        self.yaw -= delta_x * 0.01
        self.pitch = max(-1.5, min(1.5, self.pitch + delta_y * 0.01))

    def zoom_limits(self) -> tuple[float, float]:
        """Distance envelope, widened (never shrunk) so the working scale and
        the meter-scale defaults both stay reachable."""
        return (
            0.5 * min(1.0, self.view_scale),
            200.0 * max(1.0, self.view_scale),
        )

    def zoom_by(self, wheel_delta: float) -> None:
        minimum, maximum = self.zoom_limits()
        self.distance = max(
            minimum,
            min(maximum, self.distance * math.exp(-wheel_delta * 0.0012)),
        )

    def fly_step(self) -> float:
        return max(self.distance * 0.06, 0.05 * self.view_scale)

    # -- framing ---------------------------------------------------------------

    def frame_target(
        self,
        target: tuple[float, float, float],
        distance: float,
    ) -> None:
        self.target = np.array(target, dtype=np.float64)
        self.distance = float(distance)

    def frame_box(self, box) -> None:
        cx = (box.x_min + box.x_max) * 0.5
        cy = (box.y_min + box.y_max) * 0.5
        cz = (box.z_min + box.z_max) * 0.5
        extent = max(box.x_max - box.x_min, box.y_max - box.y_min,
                     box.z_max - box.z_min, 1e-3)
        self.target = np.array([cx, cy, cz], dtype=np.float64)
        self.distance = max(extent * 1.6, min(1.0, self.view_scale))

    def reframe_to_working_scale(self) -> None:
        """Reproduce the startup framing at the working scale, keeping the
        target (full workspace rescale on unit switch)."""
        self.distance = DEFAULT_VIEW_DISTANCE * self.view_scale

    # -- rays ------------------------------------------------------------------

    def screen_ray(
        self,
        x: float,
        y: float,
        width: float,
        height: float,
    ) -> tuple[np.ndarray, np.ndarray]:
        """(origin, direction) of the camera ray through screen (x, y),
        matching the QRhi surface/grid camera convention."""
        position, forward, right, up = self.basis()
        w, h = max(width, 1), max(height, 1)
        suvx = (x - 0.5 * w) / h
        suvy = -((y - 0.5 * h) / h)
        direction = 2.0 * suvx * right + 2.0 * suvy * up + self.focal * forward
        direction /= max(np.linalg.norm(direction), 1e-9)
        return position, direction

File: casoWASM/render/src/camera.rs
//! Orbit camera for the viewport, ported from `app/viewport/camera.py` plus
//! the projection matrix from the QRhi surface renderer (adapted to wgpu's
//! y-up NDC and 0..1 clip depth). The CPU `screen_ray` reproduces the GPU
//! ray exactly — keeping them bit-identical is a design invariant.

use caso_kernel::bbox::BoundingBox3D;
use caso_kernel::vec3::{vec3, Vec3};

/// The familiar startup framing: 6 m away from a 1 m grid.
pub const DEFAULT_VIEW_DISTANCE: f64 = 6.0;
const WORLD_UP: Vec3 = vec3(0.0, 0.0, 1.0);
const FALLBACK_UP: Vec3 = vec3(0.0, 1.0, 0.0);

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct OrbitCamera {
    pub target: Vec3,
    pub distance: f64,
    pub yaw: f64,
    pub pitch: f64,
    pub focal: f64,
    /// Meters per working unit; zoom floors/ceilings scale with it.
    pub view_scale: f64,
}

impl Default for OrbitCamera {
    fn default() -> Self {
        Self {
            target: Vec3::ZERO,
            distance: DEFAULT_VIEW_DISTANCE,
            yaw: 35.0_f64.to_radians(),
            pitch: 28.0_f64.to_radians(),
            focal: 1.5,
            view_scale: 1.0,
        }
    }
}

pub struct CameraBasis {
    pub position: Vec3,
    pub forward: Vec3,
    pub right: Vec3,
    pub up: Vec3,
}

impl OrbitCamera {
    pub fn position(&self) -> Vec3 {
        let cos_pitch = self.pitch.cos();
        let offset = vec3(
            cos_pitch * self.yaw.cos(),
            cos_pitch * self.yaw.sin(),
            self.pitch.sin(),
        );
        self.target + offset * self.distance
    }

    /// (position, forward, right, up), matching the shader convention.
    pub fn basis(&self) -> CameraBasis {
        let position = self.position();
        let mut forward = self.target - position;
        forward = forward / forward.length().max(1.0e-9);
        let world_up = if forward.dot(WORLD_UP).abs() > 0.99 {
            FALLBACK_UP
        } else {
            WORLD_UP
        };
        let mut right = forward.cross(world_up);
        right = right / right.length().max(1.0e-9);
        let up = right.cross(forward);
        CameraBasis {
            position,
            forward,
            right,
            up,
        }
    }

    pub fn orbit(&mut self, delta_x: f64, delta_y: f64) {
        self.yaw -= delta_x * 0.01;
        self.pitch = (self.pitch + delta_y * 0.01).clamp(-1.5, 1.5);
    }

    /// Distance envelope, widened so working scale and meter scale stay
    /// reachable.
    pub fn zoom_limits(&self) -> (f64, f64) {
        (
            0.5 * self.view_scale.min(1.0),
            200.0 * self.view_scale.max(1.0),
        )
    }

    pub fn zoom_by(&mut self, wheel_delta: f64) {
        let (minimum, maximum) = self.zoom_limits();
        self.distance = (self.distance * (-wheel_delta * 0.0012).exp()).clamp(minimum, maximum);
    }

    pub fn fly_step(&self) -> f64 {
        (self.distance * 0.06).max(0.05 * self.view_scale)
    }

    /// Pan the target in the camera plane by a screen-pixel delta.
    pub fn pan(&mut self, delta_x: f64, delta_y: f64, viewport_height: f64) {
        let basis = self.basis();
        let world_per_pixel = 2.0 * self.distance / (self.focal * viewport_height.max(1.0));
        self.target =
            self.target - basis.right * (delta_x * world_per_pixel) + basis.up * (delta_y * world_per_pixel);
    }

    pub fn frame_target(&mut self, target: Vec3, distance: f64) {
        self.target = target;
        self.distance = distance;
    }

    pub fn frame_box(&mut self, bounds: &BoundingBox3D) {
        let center = bounds.center();
        let extent = (bounds.x_max - bounds.x_min)
            .max(bounds.y_max - bounds.y_min)
            .max(bounds.z_max - bounds.z_min)
            .max(1.0e-3);
        self.target = center;
        self.distance = (extent * 1.6).max(self.view_scale.min(1.0));
    }

    /// Reproduce the startup framing at the working scale, keeping the target.
    pub fn reframe_to_working_scale(&mut self) {
        self.distance = DEFAULT_VIEW_DISTANCE * self.view_scale;
    }

    /// (origin, direction) of the camera ray through screen (x, y), matching
    /// the grid shader exactly.
    pub fn screen_ray(&self, x: f64, y: f64, width: f64, height: f64) -> (Vec3, Vec3) {
        let basis = self.basis();
        let w = width.max(1.0);
        let h = height.max(1.0);
        let suvx = (x - 0.5 * w) / h;
        let suvy = -((y - 0.5 * h) / h);
        let mut direction =
            basis.right * (2.0 * suvx) + basis.up * (2.0 * suvy) + basis.forward * self.focal;
        direction = direction / direction.length().max(1.0e-9);
        (basis.position, direction)
    }

    /// Column-major projection*view matrix for wgpu (y-up NDC, 0..1 depth).
    pub fn matrix(&self, width: u32, height: u32) -> [f32; 16] {
        let basis = self.basis();
        let eye = basis.position;
        let distance = (eye - self.target).length().max(0.1);
        let near = (distance / 1000.0).max(0.001);
        let far = (distance * 100.0).max(100.0);
        let aspect_scale = height as f64 / (width as f64).max(1.0);
        let depth_scale = far / (far - near);
        let depth_bias = -(far * near) / (far - near);

        let row0 = basis.right * (self.focal * aspect_scale);
        let row1 = basis.up * self.focal;
        let row2 = basis.forward * depth_scale;
        let row3 = basis.forward;
        let t0 = -row0.dot(eye);
        let t1 = -row1.dot(eye);
        let t2 = depth_bias - depth_scale * basis.forward.dot(eye);
        let t3 = -basis.forward.dot(eye);
        // Rows above are the matrix rows; emit column-major for WGSL.
        [
            row0.x as f32, row1.x as f32, row2.x as f32, row3.x as f32,
            row0.y as f32, row1.y as f32, row2.y as f32, row3.y as f32,
            row0.z as f32, row1.z as f32, row2.z as f32, row3.z as f32,
            t0 as f32, t1 as f32, t2 as f32, t3 as f32,
        ]
    }
}
